### PR TITLE
amazon translate

### PR DESCRIPTION
**Pull request by veronikalaskova**
Add AWS Amazon translate as another translation service option. 
Follows the same setup as other translation services.

### CODE REVIEW

File: lib/service/translation/amazon_translate.dart [hunk 1]
```diff
@@ -0,0 +1,314 @@
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:convert/convert.dart';
+import 'package:flutter_arb_translator/models/api_result.dart';
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:5]: please use relative paths for in-project references:
`import '../../models/api_result.dart';`

File: lib/service/translation/amazon_translate.dart [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+import 'package:flutter_arb_translator/service/log/logger.dart';
+import 'package:intl/intl.dart';
+
+import 'base.dart';
+
+/// Translator based on Amazon translate API
+/// Amazon Translate API Documentation:
+/// https://docs.aws.amazon.com/translate/latest/APIReference/welcome.html
+class AmazonTranslateService extends AbstractTranslationService with SupportsBulkTranslationToSingleTarget {
+  final AmazonTranslateClient client;
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:17]: I suppose `AmazonTranslateClient` should be private `_AmazonTranslateClient`

File: lib/service/translation/amazon_translate.dart [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+      sourceLanguageCode: sourceLanguage,
+      targetLanguageCode: target,
+    );
+
+    if (!apiResult.succeeded) {
+      logger.warning('Translation failed');
+      return source;
+    }
+
+    return apiResult.value!.translations.first.translatedText;
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:61]: please use `apiResult.valueUnsafe`

File: lib/service/translation/amazon_translate.dart [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+      sourceLanguageCode: sourceLanguage,
+      targetLanguageCode: target,
+    );
+
+    if (!apiResult.succeeded) {
+      logger.warning('Translation failed');
+      return sources;
+    }
+
+    return apiResult.value!.translations.map((e) => e.translatedText).toList();
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:88]: please use `apiResult.valueUnsafe`

File: lib/service/translation/amazon_translate.dart [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+/// Client for Amazon Translate API
+/// Generates signed request and sends it to Amazon Translate API
+/// [accessKeyId] - AWS Access Key Id, read mode here:
+/// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html
+/// [secretAccessKey] - AWS Secret Access Key, read mode here:
+/// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html
+/// [region] - AWS Service Endpoint Region, read more here:
+/// https://docs.aws.amazon.com/general/latest/gr/rande.html
+/// [service] - AWS Service Name, in this case it is "translate"
+class AmazonTranslateClient {
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:101]: I suppose this class should be private `_AmazonTranslateClient`

File: lib/service/translation/amazon_translate.dart [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+  final String service;
+
+  AmazonTranslateClient._({
+    required this.accessKeyId,
+    required this.secretAccessKey,
+    required this.region,
+    required this.service,
+  });
+
+  factory AmazonTranslateClient.translate({
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:114]: Shouldn't this factory method be called `create()`?

File: lib/service/translation/amazon_translate.dart [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+  String get signedHeaders => 'content-type;host;x-amz-date;x-amz-target';
+
+  String get amzDateTime => DateFormat("yyyyMMddTHHmm00'Z'").format(DateTime.now().toUtc());
+
+  String get date => DateFormat("yyyyMMdd").format(DateTime.now().toUtc());
+
+  String get algorithm => 'AWS4-HMAC-SHA256';
+
+  HttpClient buildClient(Logger logger, String message) {
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:144]: Please update visibility of methods in this class
It looks like the only public method is `translate()`. If so, please modify other methods to be private, for example `buildClient()` -> `_buildClient()`, etc

File: lib/service/translation/amazon_translate.dart [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+  Future<ApiResult<_AmazonTranslation>> translate({
+    required Logger logger,
+    required List<String> messages,
+    required String sourceLanguageCode,
+    required String targetLanguageCode,
+  }) async {
+    logger.info('Translate ${messages.length} texts from $sourceLanguageCode to $targetLanguageCode');
+
+    print("translating messages: " + messages.join("\n"));
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:242]: please remove this message or use logger

File: lib/service/translation/amazon_translate.dart [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+  }
+
+  _AmazonTranslation _decodeJson(Logger logger, String json) {
+    final String utf8 = Utf8Decoder().convert(json.codeUnits);
+
+    return _AmazonTranslation.fromJson(jsonDecode(utf8));
+  }
+}
+
+class SourceData {
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:267]: This class might be private

File: lib/service/translation/amazon_translate.dart [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    final String translationsJson = json['TranslatedText'] as String;
+    final translations = translationsJson
+        .split("\n")
+        .map((x) => AmazonTranslationItem(x.replaceAll('|', '\n')))
+        .toList();
+    return _AmazonTranslation(translations);
+  }
+}
+
+class AmazonTranslationItem {
```
Comment (sxtfv) [inline lib/service/translation/amazon_translate.dart:304]: This class might be private as well

**Review by sxtfv — CHANGES_REQUESTED** [review]
Thank you for adding Amazon translation to this project. I left some comments, which are mostly related to code formatting
Please update your PR and I will merge it

**Comment by veronikalaskova** [comment]
Hi there, 
Thanks for the quick reply.
I've pushed a commit with the requested changes.

**Review by sxtfv — APPROVED** [review]
Thanks for updating the PR

**Pull request merged**